apps/dashr-wind-steaming/app.R

```diff
@@ -0,0 +1,443 @@
+        )
+      )
+    return(fig)
+  }
+)
+
+app$callback(
+  output = list(id = "wind-direction", property = "figure"),
+  params = list(input(id = "wind-speed-update", property = "n_intervals")),
+  
+  function(interval){
+    # Generate the wind direction graph.
+    # params interval: update the graph based on an interval
+    
+    total_time <- get_current_time()
+    df <- get_wind_data_by_id(total_time)
+    val <- tail(df$Speed, n=1)
+    direction <- list(0, (df$Direction[1] - 20), (df$Direction[1] + 20), 0)
+    
+    fig <- plot_ly(
+      type = "scatterpolar",
+      mode = "lines",
+      height = 350
+    ) %>%
+      add_trace(
+        r = list(0, val, val, 0),
+        theta = direction,
+        fillcolor = "#084E8A",
+        fill = "toself",
+        line = list(color = "rgba(32, 32, 32, .6)", "width" = 1)
+      ) %>%
+      add_trace(
+        r = list(0, val * 0.65, val * 0.65, 0),
+        theta = direction,
+        fillcolor = "#B4E1FA",
+        fill = "toself",
+        line = list(color = "rgba(32, 32, 32, .6)", "width" = 1)
+      ) %>%
+      add_trace(
+        r = list(0, val * 0.3, val * 0.3, 0),
+        theta = direction,
+        fillcolor = "#EBF5FA",
+        fill = "toself",
+        line = list(color = "rgba(32, 32, 32, .6)", "width" = 1)
+      ) %>% 
+      layout(
+        plot_bgcolor = app_color[["graph_bg"]],
+        paper_bgcolor=app_color[["graph_bg"]],
+        font = list("color" = "#fff"),
+        autosize = FALSE,
+        polar = list(
+          "bgcolor" = app_color[["graph_line"]],
+          "radialaxis" = list("range" = c(0, 45), "angle" = 45, "dtick" = 10),
+          "angularaxis" = list("showline" = FALSE, "tickcolor" = "white")
+        ),
+        showlegend = FALSE
+      )
+    return(fig)
+  }
+)
+
+app$callback(
+  output = list(id = "wind-histogram", property = "figure"),
+  params = list(
+    input(id = "wind-speed-update", property = "n_intervals"),
+    list(state(id = "wind-speed", property = "figure"),
+         state(id = "bin-slider", property = "value"),
+         state(id = "bin-auto", property = "values"))
+  ),
+  
+  function(interval, wind_speed_figure, slider_value, auto_state) {
+
+    try(
+      expr = {
+        if(!is.null(wind_speed_figure)){
+          wind_val <- wind_speed_figure[['x']]$attrs[[1]]$y 
+        }
+        if("Auto" %in% auto_state){
+          bin_val <- hist(
+            wind_val,
+            breaks = round(max(unlist(wind_val)))
+          )
+        }else{
+          bin_val <- hist(wind_val, breaks = slider_value)
+        }
+      },
+      error = function(e){
+        e <- suppressWarnings()
+      }
+    )
+  
+    avg_val <- as.numeric(sum(wind_val)) / length(wind_val)
+    median_val <- median(unlist(wind_val))
+    
+    pdf_fitted <- drayleigh(bin_val[[4]], ((tail(bin_val[[1]], n= 1) - bin_val[[1]][1]) / 3))
+    y_val <- c(0, (pdf_fitted * max(bin_val[[2]]) * 20))
```

Comment:
Recommend using `max` with `na.rm=TRUE` also, unless you really want it to return `NA` if one element of your vector is missing.